Iim tests

AminKolah opened the pull request:
<!--
This template should be included in all pull requests. Items in the list should
either be completed by the original author or explicitly dismissed by one of the
IBAMR principal developers.

IBAMR is a community effort and it wouldn't exist without people contributing
code. Thanks in advance for helping to make IBAMR better!
-->

### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
- [ ] Was clang-format run?
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [ ] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

drwells committed:
Implement IIMethod

drwells committed:
Implement IIMethod

drwells committed:
Fix compilation

drwells committed:
Fix compilation

drwells committed:
TODO - undo this

drwells committed:
TODO - undo this

drwells committed:
Equivalent of 5b7c093 for IIMethod

drwells committed:
Equivalent of 5b7c093 for IIMethod

drwells committed:
Another fix

drwells committed:
Another fix

drwells committed:
run clang-format

drwells committed:
run clang-format

drwells committed:
fix end-of-function comments

drwells committed:
fix end-of-function comments

drwells committed:
Add a missing CMakeLists.txt file.

drwells committed:
Add a missing CMakeLists.txt file.

drwells committed:
Fix some warnings in IBFE-ex12.

drwells committed:
Fix some warnings in IBFE-ex12.

drwells committed:
Remove some more raw pointers.

AminKolah committed:
Adding test files for IIM

AminKolah committed:
Adding test files for IIM

AminKolah committed:
Forgot to add config files

AminKolah committed:
Forgot to add config files

AminKolah committed:
Fixing a few errros in Makefile.in

AminKolah committed:
Fixing a few errros in Makefile.in

AminKolah committed:
regenerated IIM/Makefile.in with automake

AminKolah committed:
regenerated IIM/Makefile.in with automake

AminKolah committed:
make tests complied okay for IIM examples

AminKolah committed:
make tests complied okay for IIM examples

AminKolah committed:
Further cleaning of the main files

AminKolah committed:
Further cleaning of the main files

AminKolah committed:
To be consistent in naming with the rest of the lib

AminKolah committed:
To be consistent in naming with the rest of the lib

AminKolah committed:
Changed the flow past sphere to match the explicit forcing in the paper

AminKolah committed:
Changed the flow past sphere to match the explicit forcing in the paper

AminKolah commented:
@drwells @boyceg would you mind looking at the tests?

drwells commented:
Great! I am recompiling now and will try to run the tests.

AminKolah commented:
Thanks David.

drwells commented:
They compile fine. There are some things that need to be fixed for them to work under CMake but I can handle that in a followup (unless you want to do it).

Two things:
1. The tests run really fast which is good. My goal is to keep each test under a minute and these each execute in a few seconds each. If doing more time steps would allow us to have a more meaningful test then we have the space to do it - up to you.
2. We need test output files - could you add them?

If you add corresponding output files you should be able to run the tests with `attest`.

AminKolah committed:
Added output files

AminKolah committed:
Added output files

AminKolah commented:
> They compile fine. There are some things that need to be fixed for them to work under CMake but I can handle that in a followup (unless you want to do it).
> 
> Two things:
> 
>     1. The tests run really fast which is good. My goal is to keep each test under a minute and these each execute in a few seconds each. If doing more time steps would allow us to have a more meaningful test then we have the space to do it - up to you.

Running until the full time (to evaluate convergence) for these tests would normally take a while, unless they are run with 8-16 cores but then again my laptop currently only has up to 4 threats (2 cores) so I had to set the end_time to 10*DT to keep the tests under a minute. The actual times are available in the original input files of the examples and can be separately added to IBAMR examples directory.
> 
>     2. We need test output files - could you add them?

Sure. I believe I just added them?
> 
> 
> If you add corresponding output files you should be able to run the tests with `attest`.

drwells committed:
Remove calls to boost::math::round().

We can use std::round() here instead.

drwells committed:
Merge pull request #1357 from drwells/remove-raw-pointers

Remove some more raw pointers.

drwells committed:
Merge pull request #1361 from drwells/remove-boost-round

Remove calls to boost::math::round().

drwells commented:
Sorry it took me so long to get back to this. I want to get this done this month so I will throw a lot more time at it now.

The output files are there but the tests fail at the diff stage. In particular, it looks like you generated the output files with a tighter stokes tolerance than the default (I believe you used 1e-10 while the default is 1e-6), so the logged tolerance is not right (we need to turn off logging solver details too but that can wait).

The output files are missing application-specific stuff - e.g., the flow past a sphere test saves files with the tractions and forces, which we should combine into the single output stream. Can you do this?

AminKolah commented:
Okay I see what you mean. I'll fix these by tomorrow.

drwells commented:
@AminKolah let me know if you need a hand!

AminKolah commented:
On it now..

AminKolah committed:
Fixed a couple of things based on David's suggestions

AminKolah committed:
Fixed a couple of things based on David's suggestions

AminKolah commented:
@drwells I believe I just fixed those issues..Sorry for the delay. Let me know if there's any other thing. I can regularly dedicate some time in the evening for this

drwells commented on tests/IIM/flow_past_cylinder.mpirun=2.output:
```diff
@@ -0,0 +1 @@
+../../../source/IBAMR/tests/IIM/flow_past_cylinder.mpirun=2.output
```
Adding these things as symbolic links in this manner isn't going to work - you need to actually add the files here. The test runner will take care of setting things up appropriately in the build directory.

drwells commented:
`poiseuille_flow.mpirun=2.output` has a bunch of NUL characters in it - please only use ASCII.

The `hagen_poiseuille_flow.mpirun=2.output` file doesn't match what my machine generates - in particular it looks like one wrote the ~output file~ input file and the other did not. Can you try rerunning these with `attest` and seeing what happens?

Edit: I meant to write 'input file'

AminKolah committed:
Will have to see why there are NUL chars in the outputs

AminKolah commented:
I fixed the issue with the symbolic link but still don't get where those NUL characters are coming from in the output files....Also running attest gives me this error (I get the same error in both linux and mac):

`Traceback (most recent call last):
  File "./attest", line 631, in <module>
    main()
  File "./attest", line 468, in main
    default=conf['test_timeout'],
  File "/usr/lib/python3.6/configparser.py", line 1233, in __getitem__
    raise KeyError(key)
KeyError: 'test_timeout'`

drwells commented:
That's related to a bug we fixed back in July - try rebasing, deleting the build directory, and then recompiling IBAMR. That should fix it.

AminKolah commented:
> That's related to a bug we fixed back in July - try rebasing, deleting the build directory, and then recompiling IBAMR. That should fix it.

Doesn't `drwells/iim-1` need to be rebased with `master` too? (note that my branch was off of your previous iim-1 branch and not master)...

drwells commented:
Yes it does - good catch. I rebased and force-pushed. Things should now be OK if you rebase onto `iim-1`.

AminKolah committed:
resolved the current conflicts

AminKolah commented:
> Yes it does - good catch. I rebased and force-pushed. Things should now be OK if you rebase onto `iim-1`.

Okay I'm going to try that now. Hopefully commits history won't get messed up!

AminKolah commented:
Okay I suppose we can also continue working directly on a PR of `iim-tests` against the `master` branch instead of comparing against `iim-1`, unless you find it more tractable to review the files in the current PR.

drwells commented:
We need to add these tests to `iim-1` so that we can merge the first part of IIM (that branch). We should not add these changes to any other branch.

drwells commented:
@AminKolah did you figure it out?

AminKolah commented:
My branch is already rebased against yours (and up-to-date with `master`?) I don't see any issue with the commit history but can you let me know if there is anything wrong?....Also, I'm still getting that error running with `attest`. I'm building the `iim-1` now to see if I get the same error on my macbook

AminKolah commented:
I seem to be getting the same issue even running the `master` branch with `attest`

```
Traceback (most recent call last):
  File "/Users/ekolahdouz/sfw/bigsur/ibamr/source/IBAMR/./attest", line 634, in <module>
    main()
  File "/Users/ekolahdouz/sfw/bigsur/ibamr/source/IBAMR/./attest", line 471, in main
    default=conf['test_timeout'],
  File "/usr/local/Cellar/python@3.9/3.9.7_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/configparser.py", line 1257, in __getitem__
    raise KeyError(key)
KeyError: 'test_timeout'
```

drwells commented:
Try deleting the build directory and recompiling.

AminKolah commented:
I was able to run `attest` by adding flags for `numdiff` and `mpiexec`. I guess they were missing in my original build. I'm now testing to see where the tests fail.

AminKolah committed:
A couple of tests still fail

AminKolah committed:
All iim tests passed successfully

AminKolah commented:
@drwells I just made a push. All iim tests passed successfully with `attest`. Would you mind double-checking everything looks okay? Thanks!

drwells commented:
We cannot merge until we fix the merge conflict in `configure`.

Otherwise - this looks better. I see problem-specific output for some (like `taylor_couette_2d.mpirun=2.output`) but not other (like `flow_past_sphere.mpirun=4.output`) test files. Did you mean to add more test output for those files?

AminKolah commented:
Are you referring to *.curve outputs? I think a few of the problems (flow past a cylinder/sphere) have those additional outputs (I noticed similar things in tests for IBFE ex0 and 1)... As for the configure, I see that here but I'm not seeing the conflict when I do `git pull`. Why is that? Here the option to "resolve the conflicts"  is grayed out for me saying "only those with write access to this repository can merge pull requests"

drwells commented:
> Are you referring to *.curve outputs? I think a few of the problems (flow past a cylinder/sphere) have those additional outputs (I noticed similar things in tests for IBFE ex0 and 1)...

Yes. It would be great if we could just redirect those streams to the log file so that we can verify that we get the correct values in the test suite itself.

> As for the configure, I see that here but I'm not seeing the conflict when I do git pull. Why is that?

The issue here is that, to deal with changes to IBAMR, I rebased `iim-1`. However, you did something different (you did not rebase your branch against the new `iim-1`), so you have an incompatible set of patches corresponding to where `iim-1` was in July. I'm working on a fix now.

AminKolah commented:
> > Are you referring to *.curve outputs? I think a few of the problems (flow past a cylinder/sphere) have those additional outputs (I noticed similar things in tests for IBFE ex0 and 1)...
> 
> Yes. It would be great if we could just redirect those streams to the log file so that we can verify that we get the correct values in the test suite itself.

Got it. I think I can fix this tonight, at least in my IIM tests. 
> 
> > As for the configure, I see that here but I'm not seeing the conflict when I do git pull. Why is that?
> 
> The issue here is that, to deal with changes to IBAMR, I rebased `iim-1`. However, you did something different (you did not rebase your branch against the new `iim-1`), so you have an incompatible set of patches corresponding to where `iim-1` was in July. I'm working on a fix now.

That makes sense. Sorry, I think I acted too soon when I saw `iim-1` wasn't up-to-date with IBAMR `master`. I rebased against my own `master` which was recently updated itself.

drwells commented:
I think I may have incorrectly asked that you rebase against master at one point too. Either way - the fix is straightforward. I used good old `git diff --patch` and deleted all of the changes already in IBAMR. The attached patch file contains all the relevant IIM changes on this branch:
[iim-tests.txt](https://github.com/drwells/IBAMR/files/7429258/iim-tests.txt)

you should be able to now set up a new branch with all the relevant changes by doing
```sh
> git checkout iim-1 # make sure that this is up to date
> git checkout -b iim-tests-2
> patch -p1 < iim-tests.txt
> git add tests/IIM/*
> git add tests/CMakeLists.txt tests/Makefile.am tests/Makefile.in
> git add configure configure.ac
> git commit -m "combined commits"
```
or, of course, completely rewrite this branch destructively (up to you):
```sh
> git checkout iim-1 # make sure that this is up to date
> git branch -D iim-tests
> git checkout -b iim-tests
> patch -p1 < iim-tests.txt
> git add tests/IIM/*
> git add tests/CMakeLists.txt tests/Makefile.am tests/Makefile.in
> git add configure configure.ac
> git commit -m "combined commits"
> git push -f origin iim-tests
```
and then staging all the new stuff in `tests/IIM/`. I can also set this up as a new branch if you would prefer.

AminKolah commented:
Okay thanks. I did this following the first method, except that `iim-tests` was originally on my cloned version of the library so first I had to bring in your repo with `git remote add` (was I even supposed to be able to to push anything to your repo?)  ..Anyways the new branch is now under `iim-tests-2`.

AminKolah commented:
To be continued in iim tests #3

drwells commented:
Closed in favor of #3.
